### PR TITLE
Add UPE

### PR DESCRIPTION
**Pull request by diegocurbelo**
# Changes proposed in this Pull Request:

Adds a new gateway for Stripe UPE.
- Add a feature flag for UPE
- Allow turning UPE on and off, hiding and showing the LPM gateways depending if UPE is enabled, and turning the individual UPE methods on and off.
- Migrate the UPE javascript for classic checkout from normal js asset + minified, to use the wp-scripts pipeline (webpack)

It's composed of the following PRs:
- #1654
- #1647
- #1685

# Testing instructions
- Enable UPE feature flag:
```sql
INSERT INTO `wp_options` (`option_id`, `option_name`, `option_value`, `autoload`) VALUES (NULL, '_wcstripe_feature_upe', '1', 'no') ON DUPLICATE KEY UPDATE `option_value`=VALUES(`option_value`);
```
See each PR (#1654, #1647) specific testing instructions

### CODE REVIEW

File: client/api/index.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+/* global Stripe */
```
Comment (reykjalin) [inline client/api/index.js:1]: I'm pretty sure Stripe has an npm package. Should we be using that instead?

Mostly just thinking out loud; I know we retrieve the Stripe JS via CDN, just wondering what the implications of using the npm package would be 🤔

File: client/api/index.js [hunk 2]
```diff
@@ -75,3 +83,101 @@ export const createOrder = ( sourceEvent, paymentRequestType ) => {
...
+	}
+
+	confirmIntent( url, savePaymentMethod ) {
+		console.error( 'TODO: Not implemented yet: confirmIntent' );
+		return true;
+	}
+
+	saveUPEAppearance( appearance ) {
+		console.error( 'TODO: Not implemented yet: saveUPEAppearance' );
+	}
```
Comment (reykjalin) [inline client/api/index.js:182]: Are we implementing this now or in a different PR? 🤔

File: includes/class-wc-stripe-intent-controller.php [hunk 3]
```diff
@@ -254,6 +256,65 @@ public function create_setup_intent() {
...
+	 * @param {int} $order_id The id of the order if intent created from Order.
+	 *
+	 * @return array
+	 */
+	public function create_payment_intent( $order_id = null ) {
+		$amount = WC()->cart->get_total( false );
+		$order  = wc_get_order( $order_id );
+		if ( is_a( $order, 'WC_Order' ) ) {
+			$amount = $order->get_total();
+		}
```
Comment (reykjalin) [inline includes/class-wc-stripe-intent-controller.php]: Are there any implications here if `$order` is a subclass of `WC_Order`, or do we want `WC_Order`s here specifically? I'm wondering if it'd be better to check if the `$order` variable has a `get_total()` function or if it the `WC_Order` check is preferred 🤔 

For clarity, this is what I mean:

```php
if ( method_exists( $order, 'get_total' ) ) {
```

See [method_exists()](https://www.php.net/manual/en/function.method-exists.php) documentation.

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	 */
+	protected $token_service;
+
+	/**
+	 * Create instance of payment method
+	 *
+	 * @param WC_Payments_Token_Service $token_service Instance of WC_Payments_Token_Service.
+	 */
+	public function __construct( $token_service ) {
+		//      $this->token_service = $token_service;
```
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-method.php:52]: Is this supposed to be commented out?

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	/**
+	 * Add payment method to user and return WC payment token
+	 *
+	 * @param WP_User $user User to get payment token from.
+	 * @param string  $payment_method_id Stripe payment method ID string.
+	 *
+	 * @return WC_Payment_Token_CC|WC_Payment_Token_WCPay_SEPA WC object for payment token.
+	 */
+	public function get_payment_token_for_user( $user, $payment_method_id ) {
+		//      return $this->token_service->add_payment_method_to_user( $payment_method_id, $user );
```
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-method.php:107]: Is this supposed to be commented out?

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	/**
+	 * Returns boolean dependent on whether payment method
+	 * will support saved payments/subscription payments
+	 *
+	 * @return bool
+	 */
+	public function is_reusable() {
+		return $this->is_reusable;
```
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-method.php:95]: `$this->is_reusable` is never set to anything. I can't help but wonder if that's a problem 😅

File: woocommerce-gateway-stripe.php [hunk 7]
```diff
@@ -153,6 +153,14 @@ public function init() {
...
 				require_once dirname( __FILE__ ) . '/includes/compat/class-wc-stripe-pre-orders-compat.php';
 				require_once dirname( __FILE__ ) . '/includes/class-wc-gateway-stripe.php';
+				require_once dirname( __FILE__ ) . '/includes/class-wc-stripe-features.php';
+				require_once dirname( __FILE__ ) . '/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php';
+				require_once dirname( __FILE__ ) . '/includes/payment-methods/class-wc-stripe-upe-payment-method.php';
+				require_once dirname( __FILE__ ) . '/includes/payment-methods/class-wc-stripe-upe-payment-method-cc.php';
+				require_once dirname( __FILE__ ) . '/includes/payment-methods/class-wc-stripe-upe-payment-method-giropay.php';
+				require_once dirname( __FILE__ ) . '/includes/payment-methods/class-wc-stripe-upe-payment-method-ideal.php';
+				require_once dirname( __FILE__ ) . '/includes/payment-methods/class-wc-stripe-upe-payment-method-bancontact.php';
+				require_once dirname( __FILE__ ) . '/includes/payment-methods/class-wc-stripe-upe-payment-method-eps.php';
```
Comment (reykjalin) [inline woocommerce-gateway-stripe.php:163]: Looking at this list makes me wonder if we should use a `includes/upe/` or `includes/payment-methods/upe/` folder. I have no specific opinions on that, just a thought 🙂

File: woocommerce-gateway-stripe.php [hunk 8]
```diff
@@ -267,26 +275,51 @@ public function plugin_row_meta( $links, $file ) {
...
 				} else {
-					$methods[] = 'WC_Gateway_Stripe';
-					$methods[] = 'WC_Gateway_Stripe_Sepa';
+					// These payment gateways will be hidden when UPE is enabled:
+					$methods[] = ( class_exists( 'WC_Subscriptions_Order' ) && function_exists( 'wcs_create_renewal_order' ) ) ? WC_Stripe_Subs_Compat::class : WC_Gateway_Stripe::class;
+					$methods[] = WC_Gateway_Stripe_Giropay::class;
+					$methods[] = WC_Gateway_Stripe_Ideal::class;
+					$methods[] = WC_Gateway_Stripe_Bancontact::class;
+					$methods[] = WC_Gateway_Stripe_Eps::class;
 				}
```
Comment (reykjalin) [inline woocommerce-gateway-stripe.php:301]: Isn't this whole `else` statement a duplicate of what's in the `if ( ! WC_Stripe_Features::is_upe_enabled() )` block above?

What's the difference between `WC_Stripe_Features::is_upe_enabled()` and `$this->is_upe_enabled()`?

File: woocommerce-gateway-stripe.php [hunk 9]
```diff
@@ -331,11 +366,106 @@ public function filter_gateway_order_admin( $sections ) {
...
 			public function gateway_settings_update( $settings, $old_settings ) {
 				if ( false === $old_settings ) {
-					$gateway  = new WC_Gateway_Stripe();
-					$fields   = $gateway->get_form_fields();
-					$defaults = array_merge( array_fill_keys( array_keys( $fields ), '' ), wp_list_pluck( $fields, 'default' ) );
-					return array_merge( $defaults, $settings );
+					$gateway      = new WC_Gateway_Stripe();
+					$fields       = $gateway->get_form_fields();
+					$old_settings = array_merge( array_fill_keys( array_keys( $fields ), '' ), wp_list_pluck( $fields, 'default' ) );
+					$settings     = array_merge( $old_settings, $settings );
```
Comment (reykjalin) [inline woocommerce-gateway-stripe.php:376]: Shouldn't we `return $settings` here? I see we had an early return here in the previous implementation 🤔

File: client/classic/upe/index.js [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+		const $upeContainer = $( '#wc-stripe-upe-element' );
+		blockUI( $upeContainer );
+
+		intentAction
+			.then( ( response ) => {
+				// I repeat, do NOT mount UPE twice.
+				if ( upeElement || paymentIntentId ) {
+					unblockUI( $upeContainer );
+					return;
+				}
```
Comment (reykjalin) [inline client/classic/upe/index.js:264]: I find it sort of weird (not too weird though tbh) that we have to check for this twice.

Is this indicative of some other issue, or is it truly "normal" to have a situation where you might be mounting the UPE element twice? It seems to indicate that we might be sending the same request twice at some point? Maybe that's something we should be handling instead of checking whether the element is mounted twice?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+	protected $payment_methods = [];
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		$this->id           = self::ID;
+		$this->method_title = __( 'Stripe UPE', 'woocommerce-gateway-stripe' );
+		/* translators: link */
+		$this->method_description = __( 'Accept debit and credit cards in 135+ currencies, methods such as Alipay, and one-touch checkout with Apple Pay.', 'woocommerce-gateway-stripe' );
```
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:76]: I'm not so sure this is all that helpful of a translators comment. Maybe it'd be better to say "this text is used in a link"? Or maybe just removed; I don't think basic text needs much of an explanation tbh.

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+		if ( ! isset( $this->payment_methods[ $payment_method_id ] ) ) {
+			return false;
+		}
+		return $this->payment_methods[ $payment_method_id ]->is_reusable();
+	}
+
+	// TODO: Actually validate.
+	public function validate_upe_checkout_experience_accepted_payments_field( $key, $value ) {
+		return $value;
+	}
```
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:325]: Is this supposed to be implemented now, or are we deliberately leaving this for later?

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+			</table>
+			<span id="wc_stripe_upe_change_notice" class="hidden">' . __( 'You must save your changes.', 'woocommerce-gateway-stripe' ) . '</span>';
+
+		return $this->generate_title_html( $key, $data );
+	}
+
+	/**
+	 * This is overloading the title type so the oauth url is only fetched if we are on the settings page.
+	 *
+	 * TODO: This is duplicate code from WC_Gateway_Stripe.
```
Comment (reykjalin) [inline includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:365]: Maybe this should say "TODO: create a common function we can use both here and in `WC_Gateway_Stripe`"? This todo isn't really actionable at the moment 🙂

**Review by reykjalin — DISMISSED** [review]
I couldn't actually test the PR because the `Enable new checkout experience` toggle is missing from the Stripe settings, even though I've enabled the feature flag in my DB 🤔

I tried removing the old plugin settings but that didn't help.

--- 

There are some other things I think we should look into, but most of them minor.

Overall the code looks good. I don't know if it's functional though, I'll try that after we've re-added the settings toggle 😛

File: client/api/index.js (hunk 1 above)
Comment (EvanDotPro) [reply client/api/index.js:1]: [Per Stripe](https://stripe.com/docs/js/including):

> Include the Stripe.js script on each page of your site—it should always be loaded directly from https://js.stripe.com, rather than included in a bundle or hosted yourself.

But then just below on the same page, they also say:

> We also provide an npm package that makes it easier to load and use Stripe.js as a module. For more information, check out the [project on GitHub](https://github.com/stripe/stripe-js).

Then on the docs for that project, it again emphasizes loading it from js.stripe.com:

> Note: To be [PCI compliant](https://stripe.com/docs/security/guide#validating-pci-compliance), you must load Stripe.js directly from https://js.stripe.com. You cannot include it in a bundle or host it yourself. This package wraps the global Stripe function provided by the Stripe.js script as an ES module.

Looking at the code for that module / npm package, it looks like it essentially just provides a wrapper that ultimately still just (lazy?) loads the actual stripe.js source from js.stripe.com.

tl;dr: Seems like it probably doesn't make any tangible difference.

File: client/api/index.js (hunk 2 above)
Comment (EvanDotPro) [reply client/api/index.js:182]: We want to get this PR merged and continue UPE dev in smaller, more discreet PR's against develop now that we have the feature flag.

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php (hunk 4 above)
Comment (EvanDotPro) [reply includes/payment-methods/class-wc-stripe-upe-payment-method.php:52]: Right now, yeah. The token service will be ported in a separate PR.

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php (hunk 5 above)
Comment (EvanDotPro) [reply includes/payment-methods/class-wc-stripe-upe-payment-method.php:107]: Same as above.

File: woocommerce-gateway-stripe.php (hunk 7 above)
Comment (EvanDotPro) [reply woocommerce-gateway-stripe.php:163]: I have no strong feelings either way.

File: woocommerce-gateway-stripe.php (hunk 8 above)
Comment (EvanDotPro) [reply woocommerce-gateway-stripe.php:301]: `WC_Stripe_Features::is_upe_enabled()` is if the feature flag is enabled or not. If the feature flag is not enabled, the user will not even see the setting to turn UPE on or off and all of the gateways will load exactly as they were before.

`$this->is_upe_enabled()` on the other hand is whether the merchant has turned UPE on in the Stripe settings. When The feature flag is enabled _and_ the merchant has toggled UPE on in settings, then a different set of gateways are loaded (UPE + any LPM's that aren't supported by UPE).

File: woocommerce-gateway-stripe.php (hunk 9 above)
Comment (EvanDotPro) [reply woocommerce-gateway-stripe.php:376]: If the UPE feature flag is not enabled, yes.

See change [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1685/files#diff-e1e5e9b442c2b5d4105edbdc3787651333c1923dd623677025f72d25796900d6R379-R381). (Edit: That change is now merged here.)

File: includes/payment-methods/class-wc-stripe-upe-payment-gateway.php (hunk 12 above)
Comment (EvanDotPro) [reply includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:325]: Deliberately leaving this for a separate PR. We do not want to continue making this branch / PR any larger.

File: client/classic/upe/index.js (hunk 10 above)
Comment (EvanDotPro) [reply client/classic/upe/index.js:264]: This comes directly from the WCPay implementation. I don't know the answer to your questions, but I was wondering the same thing when I saw it there.

**Review by EvanDotPro — APPROVED** [review]
LGTM. This should be safe to merge into develop now so development can continue in smaller PR's directly against develop. :ship:

File: woocommerce-gateway-stripe.php (hunk 7 above)
Comment (reykjalin) [reply woocommerce-gateway-stripe.php:163]: Then let's keep it as is for now 👍

File: client/classic/upe/index.js (hunk 10 above)
Comment (reykjalin) [reply client/classic/upe/index.js:264]: Just so we're not holding up this PR, could you create an issue to look into this further? Feel free to just link to this discussion with a short description of what we want to look into. I think the priority can be set to `low` 🙂

Comment (EvanDotPro) [reply client/classic/upe/index.js:264]: [Done.](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1689)

File: includes/payment-methods/class-wc-stripe-upe-payment-method.php (hunk 6 above)
Comment (EvanDotPro) [reply includes/payment-methods/class-wc-stripe-upe-payment-method.php:95]: It's set by the child class implementations in their constructor.

Comment (reykjalin) [reply includes/payment-methods/class-wc-stripe-upe-payment-method.php:95]: Ahh, makes sense. Thanks!

File: client/api/index.js (hunk 1 above)
Comment (reykjalin) [reply client/api/index.js:1]: > Seems like it probably doesn't make any tangible difference

Yeah, I don't think there is implementation-wise. We're probably fine as-is 🙂

File: woocommerce-gateway-stripe.php (hunk 8 above)
Comment (reykjalin) [reply woocommerce-gateway-stripe.php:301]: Ah, that makes the `WC_Stripe_Features` class make a bit more sense 😅 
I didn't understand the difference until it was pointed out. Maybe this would be clearer if `WC_Stripe_Features` was instead called `WC_Stripe_Feature_Flags`?

Or if the `WC_Stripe_Features::is_upe_enabled()` was renamed `WC_Stripe_Features::is_upe_flag_enabled()`?

I _was_ worried these were duplicates, but now I'm mostly worried about the readability 🤔 
I think renaming either the class or function would clear up some of the confusion I experienced here.

Comment (EvanDotPro) [reply woocommerce-gateway-stripe.php:301]: `WC_Stripe_Features` has been renamed to `WC_Stripe_Feature_Flags`.

File: includes/class-wc-stripe-intent-controller.php (hunk 3 above)
Comment (diegocurbelo) [reply includes/class-wc-stripe-intent-controller.php]: Using `method_exists( $order, 'get_total' )` makes sense.. changed.
